chronos/streams/asyncstream.nim

```diff
@@ -0,0 +1,908 @@
+        await rstream.buffer.wait()
+
+proc read*(rstream: AsyncStreamReader, n = 0): Future[seq[byte]] {.async.} =
+  ## Read all bytes (n <= 0) or exactly `n` bytes from read-only stream
+  ## ``rstream``.
+  ##
+  ## This procedure allocates buffer seq[byte] and return it as result.
+  if not rstream.running():
+    raise newAsyncStreamIncorrectError("Incorrect stream state")
+
+  if isNil(rstream.rsource):
+    try:
+      result = await read(rstream.tsource, n)
+    except:
+      raise newAsyncStreamReadError(getCurrentException())
+  else:
+    var res = newSeq[byte]()
+    while true:
+      let datalen = rstream.buffer.dataLen()
+      if rstream.state == Error:
+        raise newAsyncStreamReadError(rstream.error)
+      if datalen == 0 and rstream.atEof():
+        result = res
+        break
+
+      if datalen > 0:
+        let s = len(res)
+        let o = s + datalen
+        if n <= 0:
+          res.setLen(o)
+          rstream.buffer.copyData(addr res[s], 0, datalen)
+          rstream.buffer.shift(datalen)
+        else:
+          let left = n - s
+          if datalen >= left:
+            res.setLen(n)
+            rstream.buffer.copyData(addr res[s], 0, left)
+            rstream.buffer.shift(left)
+            result = res
+            break
+          else:
+            res.setLen(o)
+            rstream.buffer.copyData(addr res[s], 0, datalen)
+            rstream.buffer.shift(datalen)
+
+      await rstream.buffer.wait()
+
+proc consume*(rstream: AsyncStreamReader, n = -1): Future[int] {.async.} =
+  ## Consume (discard) all bytes (n <= 0) or ``n`` bytes from read-only stream
+  ## ``rstream``.
+  ##
+  ## Return number of bytes actually consumed (discarded).
+  if not rstream.running():
+    raise newAsyncStreamIncorrectError("Incorrect stream state")
+
+  if isNil(rstream.rsource):
+    try:
+      result = await consume(rstream.tsource, n)
+    except TransportLimitError:
+      raise newAsyncStreamLimitError()
+    except:
+      raise newAsyncStreamReadError(getCurrentException())
+  else:
+    var res = 0
+    while true:
+      let datalen = rstream.buffer.dataLen()
+      if rstream.state == Error:
+        raise newAsyncStreamReadError(rstream.error)
+      if datalen == 0:
+        if rstream.atEof():
+          if n <= 0:
+            result = res
+            break
+          else:
+            raise newAsyncStreamLimitError()
+      else:
+        if n <= 0:
+          res += datalen
+          rstream.buffer.shift(datalen)
+        else:
+          let left = n - res
+          if datalen >= left:
+            res += left
+            rstream.buffer.shift(left)
+            result = res
+            break
+          else:
+            res += datalen
+            rstream.buffer.shift(datalen)
+
+      await rstream.buffer.wait()
+
+proc write*(wstream: AsyncStreamWriter, pbytes: pointer,
+            nbytes: int) {.async.} =
+  ## Write sequence of bytes pointed by ``pbytes`` of length ``nbytes`` to
```

Comment:
Partial writes are problem of underlying stream or Transport. In current situation writeLoop of Transport will handle and break your 4gb buffer into chunks just because OS will not allow you to send 4GB of data with one `write()` call, it will returns that it sent only part of it and Transport will handle such situation.